text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,633 @@
+projects use `[patch]` to only write down `other-workspace-member = "0.2.3"` but
+this causes issues when crates later use git dependencies)
+
+If you never publish to crates.io, `path` is all you need. If crates eventually
+get published, though, they also need a `version` directive to know what version
+from crates.io you'll be depending on after the publication.
+
+Naturally, with a highly-interconnected workspace which may be relatively large,
+this leads to a lot of duplication very quickly. This is a lot of `path` and
+`version` directives that you've got to manage.
+
+## Duplication in crate versions
+
+A frequent pattern in Cargo workspaces which publish to crates.io is to have all
+the crate at the same semver version. These crates all move in lockstep during
+publication and get bumped at the same time.
+
+While a minor papercut this basically means that anyone and everyone who has a
+workspace of a lot of crates makes their own homebrew script for updating
+versions and managing updates/publications. It'd be quite convenient if we could
+standardize across the Rust ecosystem how to manage this information!
+
+## Duplication in crate metadata
+
+The last primary area of duplication that this proposal attempts to tackle is in
+crate metadata in the `[package]` section. This includes items such as:
+
+```toml
+[package]
+authors = []
+license = "..."
+repository = "..."
+documentation = "..."
+```
+
+These metadata directives are often duplicated amongst all crates, especially
+author/license/repository information. This is a pretty poor experience if you'e
+got to keep writing down the information in so many places!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Cargo's manifest parsing will be updated with new features to support
+deduplicating each of the areas above. While all of these new features are
+pretty small in their own right, they all add up to greatly reducing the
+overhead of managing a workspace of many crates. The list of new features in
+Cargo will look like the following:
+
+## Workspace-level Dependencies
+
+The `[workspace]` section can now have a `dependencies` section which works the
+same way as the `[dependencies]` section in `Cargo.toml`:
+
+```toml
+# in workspace's Cargo.toml
+[workspace.dependencies]
+log = "0.3.1"
+log2 = { version = "2.0.0", package = "log" }
+serde = { git = 'https://github.com/serde-rs/serde" }
+wasm-bindgen-cli = { path = "crates/cli" }
+```
+
+Each workspace member can then reference this section in the workspace with a
+new dependency directive:
+
+```toml
+# in a workspace member's Cargo.toml
+[dependencies]
+log = { workspace = true }
+```
+
+This directive indicates that the `log` dependency should be looked up from
+`workspace.dependencies` in the workspace root. You can also indicate which
+dependency should be used with:
+
+```toml
+[dependencies]
+log = { workspace = "log2" }
```

Comment:
It might make more sense to handle this via the "standard" package name adjustment?

```toml
[dependencies]
log = { workspace = true, package = "log2" }
```

Reply:
A good alternative to mention at least yeah! It may be a bit confusing though since the `log2` there is intended reference the left-hand-side key in the `[dependencies]` table, but the left-hand-name is not actually a package but rather the crate name to use via `--extern`. 

I was hoping to avoid that confusion and also make this a bit more succinct which is why I originally went with just `workspace = "log2"`. I'll be sure to write this up as an alternative though.

Reply:
Oh, I just written a comment on this, I didn't realized it was already mentioned.